src/rtc_stream.cpp

```diff
@@ -0,0 +1,418 @@
+{
+	if (!is_open())
+	{
+		ec = boost::asio::error::not_connected;
+		return 0;
+	}
+
+	std::size_t bytes_read = 0;
+
+	if (!m_incoming.empty())
+	{
+		std::size_t const copied = read_data(m_incoming.data(), m_incoming.size());
+		bytes_read += copied;
+		if (copied < m_incoming.size())
+		{
+			m_incoming.erase(m_incoming.begin(), m_incoming.begin() + std::ptrdiff_t(copied));
+			return bytes_read;
+		}
+
+		m_incoming.clear();
+	}
+
+	while (!m_read_buffer.empty() && m_incoming.empty() && !ec)
+	{
+		auto message = m_data_channel->receive();
+		if (!message) break;
+
+		std::visit(rtc::overloaded
+		{
+			[&](rtc::binary const& bin)
+			{
+				char const *data = reinterpret_cast<char const*>(bin.data());
+				std::size_t const size = bin.size();
+				std::size_t const copied = read_data(data, size);
+				bytes_read += copied;
+				if (copied < size)
+					m_incoming.assign(data + copied, data + size);
+			},
+			[&](rtc::string const&)
+			{
+				ec = errc::make_error_code(errc::bad_message);
+			}
+        }
+        , *message);
+	}
+
+	return bytes_read;
+}
+
+std::size_t rtc_stream_impl::write_some(error_code& ec)
+{
+	if (!is_open())
+	{
+		ec = boost::asio::error::not_connected;
+		return 0;
+	}
+
+	if (m_data_channel->bufferedAmount() > 0)
+	{
+		ec = boost::asio::error::would_block;
+		return 0;
+	}
+
+	std::size_t const max_message_size = m_data_channel->maxMessageSize();
+	std::size_t bytes_written = 0;
+	bool is_buffered = false;
+	while (!m_write_buffer.empty() && !is_buffered)
+	{
+		bytes_written += write_data(max_message_size, is_buffered);
+	}
+
+	return bytes_written;
+}
+
+void rtc_stream_impl::clear_read_buffers()
+{
+	m_read_buffer.clear();
+    m_read_buffer_size = 0;
+}
+
+void rtc_stream_impl::clear_write_buffers()
+{
+	m_write_buffer.clear();
+	m_write_buffer_size = 0;
+}
+
+std::size_t rtc_stream_impl::read_data(char const *data, std::size_t size)
+{
+	std::size_t bytes_read = 0;
+	auto target = m_read_buffer.begin();
+	while (target != m_read_buffer.end() && size > 0)
+	{
+		std::size_t to_copy = std::min(size, target->size());
+		std::memcpy(target->data(), data, to_copy);
+		data += to_copy;
+		size -= to_copy;
+		(*target)+= to_copy;
+		TORRENT_ASSERT(m_read_buffer_size >= to_copy);
+		m_read_buffer_size -= to_copy;
+		bytes_read += to_copy;
+		if (target->size() == 0) target = m_read_buffer.erase(target);
+	}
+	return bytes_read;
+}
+
+size_t rtc_stream_impl::write_data(std::size_t size, bool &is_buffered)
+{
+	std::size_t total = 0;
+	auto target = m_write_buffer.begin();
+	while (target != m_write_buffer.end())
+	{
+		total += target->size();
+		if (total >= size) break;
+		++target;
+	}
+
+	if (total > size)
+	{
+		TORRENT_ASSERT(target != m_write_buffer.end());
+		std::size_t const left = total - size;
+		std::size_t const to_copy = target->size() - left;
+		m_write_buffer.insert(target, boost::asio::const_buffer(target->data(), to_copy));
+		(*target) += to_copy;
```

Comment:
I think this is the same issue, you're not incrementing the iterator, you're incrementing the value it points to.
If `operator+=` doesn't work, you can use `std::advance()`

Reply:
Same, it's actually calling operator `+=` on the buffer itself.